### PR TITLE
Level-1 Trigger changes for HI Firmware

### PR DESCRIPTION
**Pull request by harmonicoscillator**
This PR contains a large set of changes for the Level-1 Trigger emulator to match the Heavy Ion Level 1 firmware, including several improvements to how the Stage-1 CaloParams are handled. 

Heavy Ion data taking will be in 75X, so this will be backported to 76X and 75X.

Eyes from the Stage-2 people would be very helpful to make sure I didn't mess up any Stage-2 code in the process.

### CODE REVIEW

**Comment by mmusich** [comment]
please test

**Comment by mulhearn** [comment]
+1

File: CondFormats/L1TObjects/interface/CaloParams.h [hunk 1]
```diff
@@ -38,7 +38,7 @@ namespace l1t {
       std::vector<unsigned> uparams_;
       std::vector<int> iparams_;
       std::vector<std::string> sparams_;
-      Node(){ type_="unspecified"; version_=0; }
+      Node(){ type_="unspecified"; version_=1; }
```
Comment (diguida) [inline CondFormats/L1TObjects/interface/CaloParams.h]: @richard-cms why this change of the default c'tor?

File: L1Trigger/L1TCalorimeter/data/centrality_extended_LUT_preRun.txt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+12 1685
+13 1863
+14 4113
+15 4197
+16 5040
+17 5142
+18 5940
+19 6060
+20 6930
+21 7070
```
Comment (diguida) [inline L1Trigger/L1TCalorimeter/data/centrality_extended_LUT_preRun.txt]: to be moved in the cms-data repo

Comment (davidlange6) [inline L1Trigger/L1TCalorimeter/data/centrality_extended_LUT_preRun.txt]: I don't think we need to move a 27 line data file unless there is a specific reason. (can be better if it was just in the python configs...)

File: CondFormats/L1TObjects/src/CaloParams.cc [hunk 3]
```diff
@@ -1,83 +1,3 @@
...
-  out << " Calibration type : " << this->jetCalibrationType() << std::endl;
-
-  out << "Sums" << std::endl;
-  for (unsigned i=0; i<etSumEtaMin_.size(); ++i) {
-    out << " EtSum" << i << " eta min     : " << this->etSumEtaMin(i) << std::endl;
-    if (etSumEtaMax_.size()>i) out << " EtSum" << i << " eta max     : " << this->etSumEtaMax(i) << std::endl;
-    if (etSumEtThreshold_.size()>i) out << " EtSum" << i << " Et thresh     : " << this->etSumEtThreshold(i) << std::endl;
-  }
-
-}
```
Comment (diguida) [inline CondFormats/L1TObjects/src/CaloParams.cc]: either you keep the `print` function (possibly overriding it in the wrapper sub-class) or you drop it for good.

File: L1Trigger/L1TCalorimeter/data/centrality_extended_LUT_preRun.txt (hunk 2 above)
Comment (diguida) [inline L1Trigger/L1TCalorimeter/data/centrality_extended_LUT_preRun.txt]: @davidlange6 well, I was thinking to keep the data folder clean and tidy.
But I second your point: at least move it in the python configs.

File: CondFormats/L1TObjects/interface/CaloParams.h (hunk 1 above)
Comment (richard-cms) [inline CondFormats/L1TObjects/interface/CaloParams.h]: This was requested by @mulhearn so that the emulator defaults to using the 'new' sets of parameters, he did not see any reason for emu to make version=0 parameters.

File: CondFormats/L1TObjects/src/CaloParams.cc (hunk 3 above)
Comment (richard-cms) [inline CondFormats/L1TObjects/src/CaloParams.cc]: Ah I see I should have either removed the reference in the header or kept this print function. The printer doesn't hurt anything, I'll put it back in momentarily.

File: L1Trigger/L1TCalorimeter/data/centrality_extended_LUT_preRun.txt (hunk 2 above)
Comment (richard-cms) [inline L1Trigger/L1TCalorimeter/data/centrality_extended_LUT_preRun.txt]: Sure, I can put this in /python .

**Comment by richard-cms** [comment]
ugh I messed up the stream printer, I'm just going to remove it completely since it has no customers.

**Comment by richard-cms** [comment]
We noticed that a patch for the RCT unpacker which was supposed to be pushed sometime in August never actually made it in. I have added that patch to these sets of PRs so that our HF Minimum Bias trigger emulator will function.

@ginnocen was the original author of the patch.

File: CondFormats/L1TObjects/src/CaloParams.cc (hunk 3 above)
Comment (richard-cms) [inline CondFormats/L1TObjects/src/CaloParams.cc]: Instead of putting it back in I just removed it, since I removed all the getters/setters anyway.

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
+1

**Comment by mulhearn** [comment]
@ginnocen:  I believe richard has satisfied your request, correct?

**Comment by richard-cms** [comment]
I think @mulhearn meant @diguida .

**Comment by davidlange6** [comment]
@richard-cms  - there are a lot of couts - are these in test code? (its not obvious..)

**Comment by mulhearn** [comment]
I believe these are all protected by verbose flags, but they should probably all be commented out in the version used in the release.  @richard-cms can you comment these out?

**Comment by richard-cms** [comment]
All the couts are protected by testing flags. I will comment them out for this PR.

**Comment by mulhearn** [comment]
please test

**Comment by mulhearn** [comment]
+1

File: CondFormats/L1TObjects/interface/CaloParams.h (hunk 1 above)
Comment (mulhearn) [inline CondFormats/L1TObjects/interface/CaloParams.h]: Indeed thats our model.  We'll always support reading obsolete versions, but will never write anything but the latest version.

**Comment by ggovi** [comment]
+1

**Comment by diguida** [comment]
+1
comments addressed. Code changes satisfactory

**Comment by diguida** [comment]
@richard-cms do not forget to forward port https://github.com/cms-sw/cmssw/commit/06da8c1ed9784cf974cc6672cc5bef18b21fa915 here and in #12180

**Comment by richard-cms** [comment]
@diguida  I am not sure if 06da8c1 needs to be forward-ported, that was a hotfix for a deficiency in the O2Oed CaloParams payload. If the payload has been fixed, then there is no need for the hotfix.

**Comment by mulhearn** [comment]
@kkotov can you comment on this?   Has O2O been fixed to set the jetLSB yet?

**Comment by diguida** [comment]
@richard-cms @mulhearn @kkotov thanks for looking into that.
I propose the following:
* if the O2O is ok, do not forward port
* if the O2O is not ok, cherry-pick the commit here.

**Comment by koskot77** [comment]
The working version of the O2O at p5 has the jetLSB parameter being initialized to 0 for the HI (so as to avoid a random non-physical values there). It was not yet propagated to any branch yet, but I don't think it is critical.

**Comment by diguida** [comment]
@koskot77 thanks for the quick reply.
This means that the O2O writes a value in order to keep the payload technically ok, but it is *not* the value needed for guaranteeing the optimal physics performance of the HI L1T.
If so, @richard-cms please forward port https://github.com/cms-sw/cmssw/commit/06da8c1ed9784cf974cc6672cc5bef18b21fa915 by cherry-picking it here.

**Comment by richard-cms** [comment]
@diguida agreed, will forward port 06da8c1 here.

**Comment by diguida** [comment]
please test

**Comment by mulhearn** [comment]
+1

**Comment by diguida** [comment]
+1

**Comment by diguida** [comment]
@ggovi ping :-)

**Comment by ggovi** [comment]
+1

**Comment by diguida** [comment]
@davidlange6 can you please merge this? It is required in order to integrate the HI MC GT with the new L1T menu inside, otherwise we have a SEGV.
Thanks

**Comment by davidlange6** [comment]
+1
